### ISSUE TITLE
[Id] SSR – Mismatched server / client ids when running in `StrictMode`

### ISSUE THREAD

**Issue by andy-hook** [radix-ui/primitives#811]
Related https://github.com/radix-ui/primitives/issues/793

Original report via Discord:

> I'm trying to create a Dialog with IdProvider and I can't understand why SSR doesn't work, I got an error: 
> Prop `aria-controls` did not match. Server: "radix-id-0-1" Client: "radix-id-0-4"

https://codesandbox.io/s/radixui-idprovider-ssr-mismatch-i3o1x

This is reproducible in our current SSR testing app since `StrictMode` was added in  https://github.com/radix-ui/primitives/pull/795

Wrapping `React.StrictMode` or setting `reactStrictMode: true` in `next.config` is equivalent afaik.

**Comment by andy-hook** [comment]
Thought some more about this and then it struck me that of course this would be a problem, the implementation is proven non-deterministic in concurrent mode which is why `StrictMode` is having a cry about it and incrementing twice. 

So then I thought about how might we solve for this and I kept ending up back in the same place – wouldn't it be easier to just not do this on the server in the first place? I did some digging and turns out Reach has been down this rabbit hole already and written a nice piece of documentation on it.

https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx

While this is a warning and not a "today" problem, it will become a bigger issue in the future so the question is whether we'd want to take a similar approach here / now.

Worth noting that I did check in on the react-aria implementation and it continues to suffer from the same issue.

**Comment by csantos1113** [comment]
Any updates on this? We just ran into the same issue

**Comment by andy-hook** [comment]
> Any updates on this?

Not since my follow up, one workaround is disabling `StrictMode` for now. The problem is only visible in development and production builds should behave as expected.

This issue isn't exclusive to Radix, it exists in many other `id` implementations in popular libraries. I need to speak with the team and see how we feel given the options available to us.

**Comment by csantos1113** [comment]
> This issue isn't exclusive to Radix, it exists in many other id implementations in popular libraries. I need to speak with the team and see how we feel given the options available to us.

Yeah, I created an issue in react-spectrum (https://github.com/adobe/react-spectrum/issues/2231) for react-aria to see how they react and what ideas they might have as well.

Thanks for the quick response @andy-hook - and yes, I did exactly that 😅 strictMode disabled for now

**Comment by Clee681** [comment]
Seeing the same issue was a `NextJS` user

**Comment by julianbenegas** [comment]
Any updates on this? It's a super annoying error 😅

**Comment by espired** [comment]
My logs gets filled up with this

**Comment by andy-hook** [comment]
I agree that this warning is not great but our options to resolve it without compromise are limited. In order to deal with this today we’d likely need to render id’s on the client only, this has [performance and a11y consequences](https://github.com/radix-ui/primitives/issues/331#issuecomment-760158869).

Once stable our best bet will be [useOpaqueIdentifier](https://github.com/facebook/react/pull/17322). For now disabling `StrictMode` remains the best workaround 🙏

**Comment by ashtonlance** [comment]
@andy-hook I have a vercel staging url I can share! unless you need the source code -- that repo is private but I could add you as a user.

**Comment by woshixixi** [comment]
same issue here, bug with tooltip component

**Comment by andy-hook** [comment]
> @andy-hook I have a vercel staging url I can share! unless you need the source code -- that repo is private but I could add you as a user.

@ashtonlance If you could add me that'd be great.

**Comment by albingroen** [comment]
Same issue for me, and I'm using the Dialog components.

**Comment by ashtonlance** [comment]
> > @andy-hook I have a vercel staging url I can share! unless you need the source code -- that repo is private but I could add you as a user.
> 
> @ashtonlance If you could add me that'd be great.

Added. Thanks for taking a look!

**Comment by coderinblack08** [comment]
Possible react-18 solution?
https://github.com/reactwg/react-18/discussions/111

**Comment by jjenzz** [comment]
@coderinblack08 Yep, that is the renamed `useOpaqueIdentifier` hook we mentioned. Looking forward to giving it a spin 💃

**Comment by ajrussellaudio** [comment]
For anyone late to the party dropping in from Google, not able to upgrade to React 18 yet, [`suppressHydrationWarning` (docs)](https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning) on the JSX element causing the warning worked for me.

**Comment by jjenzz** [comment]
@ajrussellaudio If you are able to upgrade to the latest versions of the primitives, the `IdProvider` is no longer needed and these warnings should all go away regardless of what React version you are on.

**Comment by ajrussellaudio** [comment]
Awesome @jjenzz thanks!

**Comment by daniellizik** [comment]
@ajrussellaudio how are you able to determine which elements require the `suppresshydrationwarning` flag? My terminal just spits out this message a bunch of times without any reference to where it originates from. i'm on react@17.0.2 and @radix/react-id@0.1.6

**Comment by ajrussellaudio** [comment]
@daniellizik By manual process of elimination, commenting out components until the error disappears and following the tree down. 

But as @jjenzz points out, this fix is no longer needed if you upgrade your radix primitives dependency.

**Comment by DennieMello** [comment]
Problem reappeared, using react: 18.2.0, next: 13.4.12 with app router, @radix-ui/react-dialog.

Error: app-index.js:31 Warning: Prop `aria-controls` did not match. Server: "radix-:R16qcq:" Client: "radix-:R4r9j9:"

Any ideas for a fix besides setting a custom value in aria-controls?

**Comment by joaom00** [comment]
@DennieMello Can you check if this is a problem on the Next.js side? https://github.com/radix-ui/primitives/issues/1684#issuecomment-1665811607

**Comment by DennieMello** [comment]
> @DennieMello Can you check if this is an issue on the Next.js side? [#1684 (comment)](https://github.com/radix-ui/primitives/issues/1684#issuecomment-1665811607)

Thanks, fixed in next: "13.4.13".

**Comment by Kinbaum** [comment]
I think I just encountered something like this with Next.js 14, app router, and dynamic imports.
I'm dynamically importing a dialog component using next/dynamic and ssr set to true so my trigger is visible on the page.

When the code for the dialog downloads, I notice my trigger disappear, and reappear.
When looking at what is rendered on the server vs the client (after the flicker), I see the following (note the aria-controls value):

```jsx
// In the browser (post flicker)
<button 
    data-state="closed" 
    type="button" 
    aria-haspopup="dialog" 
    aria-expanded="false" 
    aria-controls="radix-:rn:" 
    class="..."
    data-rac="">
    click here
</button>
```

```jsx
// What's rendered on the server
<button
    data-state="closed"
    type="button"
    aria-haspopup="dialog"
    aria-expanded="false"
    aria-controls="radix-:Rdlauujfaudta:"
    class="..."
    data-rac="">
    click here
</button>
```

Has anyone else attempted to dynamically import code from a page using the app router?
Since my page is a server component, using next/dynamic there doesn't work, the code is still part of the main bundle for that route.

What I had to do to actually get the code splitting to work was create a proxy component of sorts that has the use client directive, and let that component handle the dynamic import. Something like this:

```tsx
// app/page.tsx

import ProxyDialog from 'app/proxy-dialog';

export default async function Page() {
    // Some data fetching
    return (
        <ProxyDialog />
    )
}
```

```tsx
// app/proxy-dialog.tsx

'use client';

import dynamic from 'next/dynamic';

const MyDialog = dynamic(() => import('app/my-dialog')); // Note here that I want this to be SSR'd

export default function ProxyDialog() {
    return (
        <MyDialog />
    )
}
```
